### ISSUE TITLE
TF1.8 is 10% slower than TF1.4 on most cases

### ISSUE THREAD

**Issue by ghostplant** [tensorflow/tensorflow#19559]
I have tested on quantities of tensorflow-based applications, including tf_cnn_benchmark and keras typically applications and even for multi-GPU applications. All performance benchmarks are 10% dropped when TF1.4 is upgraded to TF1.8. However, once I revert the upgrade to use TF1.4 again, the 10% performance loss will be back.

**Comment by ghostplant** [comment]
Originally, training inceptionV3 per epoch costs 62sec, after bare upgrade from TF1.4 to TF1.8 without model code changes, fininshing 1 epoch costs 85sec.

**Comment by asimshankar** [comment]
CC @tfboyd

**Comment by tfboyd** [comment]
@ghostplant   Can you provide an example command line for tf_cnn_benchmarks?

**Comment by ghostplant** [comment]
I tested TF1.6 as well, which has a same performance with TF1.4. So there might be a performance regression between TF1.6 and TF1.8.

**Comment by tfboyd** [comment]
@ghostplant 

I know the script well.  I published our original benchmarks and run nightly tests.  Knowing the command-line and your hardware helps me understand the problem.  I do not test Keras and I think Keras multi-gpu places all of the variables on GPU:0, which is often not great but can be "ok" with small numbers of GPUs and maybe something is making that slower.  tf_cnn_benchmarks is nice as we can kind of isolate that and put the parameters almost anywhere.  Hardware is good to know as you could have anything from nvlinked V100s to K80s PCIe with no GPU peer-to-peer or two GTX1080tis in a workstation also most likely not peer-to-peer.  You could be using nccl, hierarchical copy, fp16, parameter server = cpu and one of a number of args for tf_cnn_benchmarks.  

My tests show that we are speeding up but **I do not doubt your regression**.  I know many argument combinations are not tracked, which is my fault and a result of pragmatic time use and maybe bad choices. :-)

There is light at the end of the tunnel.  While it will not help you with Keras, DistributedStrategies (the standard way of doing TensorFlow multi-GPU) will be enabled for Keras models and that will over a simple, flexible, and monitored way to do multi-GPU.

Send me a little more information and I will run some tests.  I could run some tests now but I fear they would not match your situation and just use up time.

**Comment by ghostplant** [comment]
My environment is Tesla P100x4 without NV-link. The initialization time for Keras application are all increased, and I think both CUDA environment and Keras engine itself are not the root-causes, and the example script is also not the root-cause as well, because all of them are not changed, so the only changes is the version of Tensorflow for cuda.

**Comment by ppwwyyxx** [comment]
Again, can you provide your command line for tf_cnn_benchmarks? 

There are many command line options in tf_cnn_benchmarks, and for resnet50 which I'm regularly tracking, the best performing option is getting faster since 1.4. You're probably using a different combination of options and that's why a command line is needed.

**Comment by ghostplant** [comment]
`tf_cnn_benchmarks/tf_cnn_benchmarks.py --model resnet50 --batch_size 64 --num_batches 50 --data_format=NCHW --local_parameter_device cpu --num_inter_threads 0`

Besides, I refer from the benchmark links: https://www.tensorflow.org/performance/benchmarks#details_for_nvidia_dgx-1tm_nvidia_tesla_p100

which says inceptionV3 can have a 569 images/sec using P100 x 4 (parameter on cpu), and I have the same hardware configuration but I never saw this speed (it only reaches 490 images/sec). Can you also tell me how to manage to make it 569 images/sec?

**Comment by ppwwyyxx** [comment]
Running your command on 4 P100s (with the extra argument --num_gpus=4) gave me 870 im/s. This command may not be the best but at least it's slightly better than the old numbers (852) in https://www.tensorflow.org/performance/benchmarks. So it looks alright. My environment does have nvlink (the above link does mention nvlink), btw. Note that the best combination of options would also depend on your environment. Perhaps there is a regression on environment without nvlink but more information will be needed to find out if that's true.

**Comment by ghostplant** [comment]
`--local_parameter_device cpu` doesn't require NVlink. How about your inceptionV3 performance?

**Comment by ppwwyyxx** [comment]
I got 535 im/s with the same command above except for `--model inception3`. This is a bit slower than the official number but I don't know if that's a regression -- unlike resnet50 I don't track the performance of inceptionv3 myself so I don't know what its performance was on my environment. Just to provide a data point for your reference. @tfboyd may have more information on it.

**Comment by ppwwyyxx** [comment]
I don't quite understand from above what exactly is your observation or how you obtain them. But it seems related to the fact that none of the convolutional layers in ResNet50 have bias.

**Comment by ghostplant** [comment]
@ppwwyyxx It can be profiled by any tools, typically, nvtools, syscall counters. Are you sure Resnet50 CNN layers don't include bias weight?

**Comment by ppwwyyxx** [comment]
I'm sure none of the convolutional layers in ResNet50 have bias. I'm not sure you mean the same thing by "bias weight". The convolution filters are often referred to as weight as well.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @tfboyd: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by ghostplant** [comment]
Thanks all, I'll close this issue. I think larger ops and function libraries in new versions might cost more memory which is reasonable to slightly reduce the performance.
